### PR TITLE
plan: handle type convertion when building index range.

### PR DESCRIPTION
**Pull request by coocood**
Also add cast string for like and regexp expression.

### CODE REVIEW

**Comment by hanfei1991** [comment]
LGTM

File: plan/typeinferer.go [hunk 1]
```diff
@@ -384,6 +380,48 @@ func (v *typeInferrer) handleCaseExpr(x *ast.CaseExpr) {
 	x.Type.Charset, x.Type.Collate = types.DefaultCharsetForType(x.Type.Tp)
 }
 
+// like expression expect the target expression and pattern to be a string, if it's not, we add a cast function.
```
Comment (zimulala) [inline plan/typeinferer.go]: s/expect/expects

File: plan/typeinferer.go [hunk 2]
```diff
@@ -384,6 +380,48 @@ func (v *typeInferrer) handleCaseExpr(x *ast.CaseExpr) {
...
+func (v *typeInferrer) handleLikeExpr(x *ast.PatternLikeExpr) {
+	x.SetType(types.NewFieldType(mysql.TypeLonglong))
+	x.Type.Charset = charset.CharsetBin
+	x.Type.Collate = charset.CollationBin
+	x.Expr = v.addCastToString(x.Expr)
+	x.Pattern = v.addCastToString(x.Pattern)
+}
+
+// regexp expression expect the target expression and pattern to be a string, if it's not, we add a cast function.
```
Comment (zimulala) [inline plan/typeinferer.go]: s/expect/expects

File: plan/range.go [hunk 3]
```diff
@@ -456,32 +467,85 @@ func (r *rangeBuilder) buildIndexRanges(rangePoints []rangePoint) []*IndexRange
...
+		} else {
+			if valCmpCasted > 0 {
+				// e.g "a >= 1.1 convert to "a > 1"
+				point.excl = true
+			}
+		}
+	} else {
+		if point.excl {
+			if valCmpCasted > 0 {
+				// e.g, "a < 1.1" convert to "a <= 1"
```
Comment (zimulala) [inline plan/range.go]: s/e.g,/e.g.

File: plan/typeinferer.go [hunk 4]
```diff
@@ -384,6 +380,48 @@ func (v *typeInferrer) handleCaseExpr(x *ast.CaseExpr) {
...
+		castTp := types.NewFieldType(mysql.TypeString)
+		castTp.Charset, castTp.Collate = types.DefaultCharsetForType(mysql.TypeString)
+		if val, ok := expr.(*ast.ValueExpr); ok {
+			newVal, err := val.Datum.ConvertTo(castTp)
+			if err != nil {
+				v.err = errors.Trace(err)
+			}
+			expr.SetDatum(newVal)
+		} else {
+			castFunc := &ast.FuncCastExpr{
```
Comment (zimulala) [inline plan/typeinferer.go]: s/castFunc :=/ expr =

File: plan/plan_test.go [hunk 5]
```diff
@@ -1055,60 +1108,94 @@ func (s *testPlanSuite) TestRefine(c *C) {
...
+			best: "Index(t.c_d_e_str)[[abc_,abc`)]->Projection",
 		},
 		{
-			sql:  `select a from t where c like 'abc\\__'`,
-			best: "Index(t.c_d_e)[(abc_,abc`)]->Selection->Projection",
+			sql:  `select a from t where c_str like 'abc\\__'`,
+			best: "Index(t.c_d_e_str)[(abc_,abc`)]->Selection->Projection",
+		},
+		{
+			// check that 123 is converted to string '123'. index can be used.
```
Comment (zimulala) [inline plan/plan_test.go]: s/check /Check

**Comment by zimulala** [comment]
LGTM

**Pull request merged**